lib/route/cls/mall.c

```diff
@@ -0,0 +1,288 @@
+
+	if (!(mall = rtnl_tc_data(TC_CAST(cls))))
+		return -NLE_NOMEM;
+
+	if (!(mall->m_mask & MALL_ATTR_FLAGS))
+		return -NLE_INVAL;
+
+	*flags = mall->m_flags;
+	return 0;
+}
+
+int rtnl_mall_add_action(struct rtnl_cls *cls, struct rtnl_act *act)
+{
+	struct rtnl_mall *mall;
+
+	if (!act)
+		return 0;
+
+	if (!(mall = rtnl_tc_data(TC_CAST(cls))))
+		return -NLE_NOMEM;
+
+	mall->m_mask |= MALL_ATTR_ACTION;
+	rtnl_act_get(act);
+	return rtnl_act_append(&mall->m_act, act);
+}
+
+struct rtnl_act *rtnl_mall_get_action(struct rtnl_cls *cls)
+{
+	struct rtnl_mall *mall;
+	struct rtnl_act *act;
+
+	if (!(mall = rtnl_tc_data(TC_CAST(cls))))
+		return NULL;
+
+	if (!(mall->m_mask & MALL_ATTR_ACTION))
+		return NULL;
+
+	act = mall->m_act;
+	rtnl_act_get(act);
+
+	return act;
+}
+
+int rtnl_mall_del_action(struct rtnl_cls *cls, struct rtnl_act *act)
+{
+	struct rtnl_mall *mall;
+	int ret;
+
+	if (!act)
+		return 0;
+
+	if (!(mall = rtnl_tc_data(TC_CAST(cls))))
+		return -NLE_NOMEM;
+
+	if (!(mall->m_mask & MALL_ATTR_ACTION))
+		return -NLE_INVAL;
+
+	ret = rtnl_act_remove(&mall->m_act, act);
+	if (ret)
+		return ret;
+
+	if (!mall->m_act)
+		mall->m_mask &= ~MALL_ATTR_ACTION;
+	rtnl_act_put(act);
+
+	return 0;
+}
+
+/** @} */
+
+static void mall_free_data(struct rtnl_tc *tc, void *data)
+{
+	struct rtnl_mall *mall = data;
+
+	if (mall->m_act)
+		rtnl_act_put_all(&mall->m_act);
+}
+
+static int mall_msg_parser(struct rtnl_tc *tc, void *data)
+{
+	struct rtnl_mall *mall = data;
+	struct nlattr *tb[TCA_MATCHALL_MAX + 1];
+	int err;
+
+	err = tca_parse(tb, TCA_MATCHALL_MAX, tc, mall_policy);
+	if (err < 0)
+		return err;
+
+	if (tb[TCA_MATCHALL_CLASSID]) {
+		mall->m_classid = nla_get_u32(tb[TCA_MATCHALL_CLASSID]);
+		mall->m_mask |= MALL_ATTR_CLASSID;
+	}
+
+	if (tb[TCA_MATCHALL_FLAGS]) {
+		mall->m_flags = nla_get_u32(tb[TCA_MATCHALL_FLAGS]);
+		mall->m_mask |= MALL_ATTR_FLAGS;
+	}
+
+	if (tb[TCA_MATCHALL_ACT]) {
+		mall->m_mask |= MALL_ATTR_ACTION;
+		err = rtnl_act_parse(&mall->m_act, tb[TCA_MATCHALL_ACT]);
+		if (err)
+			return err;
+	}
+
+	return 0;
+}
+
+static int mall_msg_fill(struct rtnl_tc *tc, void *data, struct nl_msg *msg)
+{
+	struct rtnl_mall *mall = data;
+
+	if (!mall)
+		return 0;
+
+	if (mall->m_mask & MALL_ATTR_CLASSID)
+		NLA_PUT_U32(msg, TCA_MATCHALL_CLASSID, mall->m_classid);
+
+	if (mall->m_mask & MALL_ATTR_FLAGS)
+		NLA_PUT_U32(msg, TCA_MATCHALL_FLAGS, mall->m_flags);
+
+	if (mall->m_mask & MALL_ATTR_ACTION) {
+		int err;
+
+		err = rtnl_act_fill(msg, TCA_MATCHALL_ACT, mall->m_act);
+		if (err)
+			return err;
+	}
+
+	return 0;
+
+ nla_put_failure:
+	return -NLE_NOMEM;
+}
+
+static int mall_clone(void *_dst, void *_src)
+{
+	struct rtnl_mall *dst = _dst, *src = _src;
+
+	if (src->m_act) {
+		if (!(dst->m_act = rtnl_act_alloc()))
+			return -NLE_NOMEM;
+
+		memcpy(dst->m_act, src->m_act, sizeof(struct rtnl_act));
+
+		/* action nl list next and prev pointers must be updated */
+		nl_init_list_head(&dst->m_act->ce_list);
```

Comment:
this only copies the first action. Is that intentional? It should deep-clone the entire list.